lib/image_composition.dart

```diff
@@ -0,0 +1,167 @@
+  /// to [defaultAntiAlias].
+  ///
+  /// The [blendMode] can be used to change how the [image] will be blended onto
+  /// the composition. Defaults to [defaultBlendMode].
+  void add(
+    Image image,
+    Vector2 position, {
+    Rect source,
+    double rotation = 0,
+    Vector2 anchor,
+    bool isAntiAlias,
+    BlendMode blendMode,
+  }) {
+    assert(image != null, 'Image is required to add to the Atlas');
+    assert(position != null, 'Position is required');
+    assert(
+      source == null ||
+          source.width <= image.width &&
+              source.height <= image.height &&
+              source.top + source.height <= image.height &&
+              source.top >= 0 &&
+              source.left + source.width <= image.width &&
+              source.left >= 0,
+      'Source rect should fit within in the image constraints',
+    );
+    assert(rotation != null, 'rotation can not be null');
+    blendMode ??= defaultBlendMode;
+    isAntiAlias ??= defaultAntiAlias;
+    source ??= Rect.fromLTWH(
+      0,
+      0,
+      image.width.toDouble(),
+      image.height.toDouble(),
+    );
+    anchor ??= Vector2(source.width / 2, source.height / 2);
+
+    _composes.add(_Composed(
+      image,
+      position,
+      source,
+      rotation,
+      anchor,
+      isAntiAlias,
+      blendMode,
+    ));
+  }
+
+  void clear() => _composes.clear();
+
+  /// Compose all the images into a single composition.
+  Future<Image> compose() async {
+    // Rect used to determine how big the output image will be.
+    var output = const Rect.fromLTWH(0, 0, 0, 0);
+    final recorder = PictureRecorder();
+    final canvas = Canvas(recorder);
+
+    for (var i = 0; i < _composes.length; i++) {
+      final image = _composes[i].image;
+      final position = _composes[i].position;
+      final source = _composes[i].source;
+      final rotation = _composes[i].rotation;
+      final anchor = _composes[i].anchor;
+      final isAntiAlias = _composes[i].isAntiAlias;
+      final blendMode = _composes[i].blendMode;
+      final destination = Rect.fromLTWH(0, 0, source.width, source.height);
+      final realDest = destination.translate(position.x, position.y);
+
+      canvas
+        ..save()
+        ..translateVector(position)
+        ..translateVector(anchor)
+        ..rotate(rotation)
+        ..translateVector(-anchor)
+        ..drawImageRect(
+          image,
+          source,
+          destination,
+          Paint()
+            ..blendMode = blendMode
+            ..isAntiAlias = isAntiAlias,
+        )
+        ..restore();
+
+      // If the realDest is not fully contained inside the current output,
+      // we have to expande the output so it includes the realDest.
+      if (!(output.contains(realDest.topLeft) &&
+          output.contains(realDest.topRight) &&
+          output.contains(realDest.bottomLeft) &&
+          output.contains(realDest.bottomRight))) {
+        output = output.expandToInclude(realDest);
```

Comment:
would expandToInclude no-op if it already contained?

Reply:
Indeed it does. :-)